lib/_http_agent.js

```diff
@@ -168,6 +168,7 @@ Agent.prototype.addRequest = function addRequest(req, options, port/*legacy*/,
     var socket = this.freeSockets[name].shift();
     // Assign the handle a new asyncId and run any init() hooks.
     socket._handle.asyncReset();
+    socket[async_id_symbol] = socket._handle.getAsyncId();
```

Comment:
For the moment I'd say this is the correct way to handle the fix, and was an oversight on my part.

@refack 
> I don't know why but my intuition was to do this somewhere in a `Socket` method. Maybe move both lines to a `Socket` method so other `Agent`s could use it. Something like `Socket.prototype._prepareForReuse()` or `_recycle`

Technically `async_id_symbol` was only added as a convenience to skip calling into native and retrieving `AsyncWrap::async_id_`, but in reality is unnecessary. The most sure way to handle this would be to remove all uses of `async_id_symbol` where there is a 1:1 pairing of native resource to JS resource and always call `getAsyncId()`.

@addaleax 
> we would want to consider what kind of API, if any, custom Agent implementations need.

Node allows custom implementation of  agents, timers and a few others. Which is the reason why `getNewAsyncId()` in `lib/net.js` checks if `getAsyncId === 'function'` and why `insert()` in `lib/timers.js` forcefully adds a new `asyncId` if none already exists. It has been a total pain during development, and the existing solution is the best I could come up with. Though if you have a better idea I would be happy to try it out.